expfmt/encode_test.go

```diff
@@ -0,0 +1,122 @@
+package expfmt
+
+import (
+	"bytes"
+	"github.com/golang/protobuf/proto"
+	dto "github.com/prometheus/client_model/go"
+	"net/http"
+	"testing"
+)
+
+func TestNegotiate(t *testing.T) {
+	acceptValuePrefix := "application/vnd.google.protobuf;proto=io.prometheus.client.MetricFamily"
+	tests := []struct {
+		name              string
+		acceptHeaderValue string
+		expectedFmt       string
+	}{
+		{
+			name:              "delimited format",
+			acceptHeaderValue: acceptValuePrefix + ";encoding=delimited",
+			expectedFmt:       string(FmtProtoDelim),
+		},
+		{
+			name:              "text format",
+			acceptHeaderValue: acceptValuePrefix + ";encoding=text",
+			expectedFmt:       string(FmtProtoText),
+		},
+		{
+			name:              "compact text format",
+			acceptHeaderValue: acceptValuePrefix + ";encoding=compact-text",
+			expectedFmt:       string(FmtProtoCompact),
+		},
+		{
+			name:              "plain text format",
+			acceptHeaderValue: "text/plain;version=0.0.4",
+			expectedFmt:       string(FmtText),
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			h := http.Header{}
+			h.Add(hdrAccept, test.acceptHeaderValue)
+			actualFmt := string(Negotiate(h))
+			if actualFmt != test.expectedFmt {
+				t.Errorf("expected Negotiate to return format %s, but got %s instead", test.expectedFmt, actualFmt)
+			}
+		})
+	}
+}
+
+func TestEncode(t *testing.T) {
+	var buff bytes.Buffer
+	delimEncoder := NewEncoder(&buff, FmtProtoDelim)
+	metric := &dto.MetricFamily{
+		Name: proto.String("foo_metric"),
+		Type: dto.MetricType_UNTYPED.Enum(),
+		Metric: []*dto.Metric{
+			{
+				Untyped: &dto.Untyped{
+					Value: proto.Float64(1.234),
+				},
+			},
+		},
+	}
+
+	err := delimEncoder.Encode(metric)
+	if err != nil {
+		t.Errorf("unexpected error during encode: %s", err.Error())
+	}
+
+	out := buff.Bytes()
+	if len(out) == 0 {
+		t.Errorf("expected the output bytes buffer to be non-empty")
```

Comment:
This could test the actual content. Same below where only non-emptiness as asserted.